## Problem #147: Rectangles in Cross-hatched Grids
[Link to Problem](https://projecteuler.net/problem=147)

### Problem Description

In a $3 \times 2$ cross-hatched grid, a total of $37$ different rectangles could be situated within that grid as indicated in the sketch.

![image](https://projecteuler.net/resources/images/0147.png?1678992052)

There are $5$ grids smaller than $3 \times 2$, vertical and horizontal dimensions being important, i.e. $1 \times 1$, $2 \times 1$, $3 \times 1$, $1 \times 2$ and $2 \times 2$. If each of them is cross-hatched, the following number of different rectangles could be situated within those smaller grids:

| $1 \times 1$  | $1$    |
| -------- | ------- |
| $2 \times 1$  | $4$    |
| $3 \times 1$  | $8$    |
| $1 \times 2$  | $4$    |
| $2 \times 2$  | $18$    |

Adding those to the $37$ of the $3 \times 2$ grid, a total of $72$ different rectangles could be situated within $3 \times 2$ and smaller grids.

How many different rectangles could be situated within $47 \times 43$ and smaller grids?

### Approach

Let $M = 47$ (number of columns of the grid) and $N = 43$ (number of rows of the grid).

The rectangles with sides parallel to the axes are easy to count; for a grid of size $M \times N$, the number of such rectangles is:

$$

\text{count}_{straight} = \sum\limits_{h=1}^{N} \sum\limits_{w=1}^{M} (N - h + 1)(M - w + 1) = \frac{1}{4}N(N+1)M(M+1)

$$

To count the diagonal rectangles it's trickier. One interesting way to count them is to imagine that each diagonal rectangle has a bounding box with the sides parallel to the axes. If we can find out how many possible rectangles can fit in a bounding box of any size, we can just iterate over the dimensions of the bounding box and thus solve the problem.

Due to the fact that all the lines are either parallel to the sides or at a $45^\circ$ angle, all the bounding boxes are actually squares! Let's add extra horizontal and vertical lines and let's look at some examples:


![Example 1](147_example_1.png)
![Example 2](147_example_2.png)
![Example 3](147_example_3.png)
![Example 4.1](147_example_4_1.png)
![Example 4.2](147_example_4_2.png)

Let  $f(k)$ be the *average* number of rectangles that can fit in a bounding box of size $k$ ($k \leq 2min(N, M)$). Why do we care about the average? Because as we see the number of rectangles varies by 1, which is troublesome, but if we can pair every two different bounding boxes and reach an average, the calculations become easy. It appears that:

$$

f(k) = \frac{k - 1}{2}

$$

For example, $f(1) = 0$, $f(2) = 0.5$, $f(3) = 1$, $f(4) = 1.5$, $f(5) = 2$. Therefore we can compute the number of diagonal rectangles based on this.

$$

\text{count}_{diagonal} = \sum\limits_{k=1}^{2min(N, M)} (2N - k + 1)(2M - k + 1)f(k)

$$

Now there is a caveat: when k is even, the result we get is actually bigger with $0.5$ than it should actually be because the first top-left corner square of size k has no pair for i. Therefore we have to substract $0.5$ from the answer for every even k.

In [15]:
M, N = 47, 43

def f(k):
    return (k - 1) / 2

total = 0

for n in range(1, N + 1):
    for m in range(1, M + 1):
        straight = n * (n + 1) * m * (m + 1) // 4
        diagonal = 0
        for k in range(1, 2 * min(n, m) + 1):
            count = (2 * n - k + 1) * (2 * m - k + 1)
            diagonal += int(count * f(k))
        total += straight + diagonal

print(total)

846910284


###### Result: **846910284** | Execution time: 0s

### Complexity analysis

Time complexity: $O(NMmin(N,M))$

##### Tags: #geometry, #counting, #math